File: Java/com/leetcode/binaryTreeMaximumPathSum/BinaryTreeMaximumPathSum.java
package com.leetcode.binaryTreeMaximumPathSum;

import com.leetcode.dataStructures.TreeNode;

public class BinaryTreeMaximumPathSum {
    public int maxPathSum(TreeNode root) {
        if (root == null){
            return 0;
        }
        if (root.left == null && root.right == null){
            return root.val;
        }
        int[] crossMax = new int[1];
        crossMax[0] = Integer.MIN_VALUE;
        int rootEndMax = getMax(root, crossMax);
        return (rootEndMax > crossMax[0] ? rootEndMax : crossMax[0]);
        
    }
    
    int getMax(TreeNode root, int[] crossMax){
        int leftMax = 0;
        int rightMax = 0 ;
        if (root.left == null && root.right == null){
            if (root.val > crossMax[0]){
                crossMax[0] = root.val;
            }
            return root.val;
        }
        if (root.left != null){
            leftMax = getMax(root.left, crossMax);    
        }
        if (root.right != null){
            rightMax = getMax(root.right, crossMax);
        }
        
        int cross = root.val;
        if (leftMax > 0){
            cross += leftMax;
        }
        
        if (rightMax > 0){
            cross += rightMax;
        }
        
        if (cross > crossMax[0]){
            crossMax[0] = cross;
        }
        
        
        return (root.val > ((leftMax + root.val) > (rightMax + root.val) ? (leftMax + root.val) : (rightMax + root.val)) ? root.val : ((leftMax + root.val) > (rightMax + root.val) ? (leftMax + root.val) : (rightMax + root.val)));
    }
}

File: Python/binary_tree_maximum_path_sum/binary_tree_maximum_path_sum.py
class Solution(object):
    def maxPathSum(self, root):
        if root is None:
            return 0
        self.MIN = -2**31
        max_val, _ = self.get_max(root)
        return max_val

    def get_max(self, root):
        if root.left is None and root.right is None:
            return root.val, root.val
        left_max, left_path_max = self.get_max(root.left) if root.left else (self.MIN, self.MIN)
        right_max, right_path_max = self.get_max(root.right) if root.right else (self.MIN, self.MIN)
        curr_path_max = max(root.val,
                                                left_path_max + root.val,
                                                right_path_max + root.val)
        curr_max = max(left_max,
                                        right_max,
                                        curr_path_max,
                                        left_path_max + right_path_max + root.val)
        return curr_max, curr_path_max
